Public/Src/Cache/ContentStore/Library/Stores/FileSystemContentStoreInternal.cs

```diff
@@ -2202,17 +2210,16 @@ public async Task<EvictResult> EvictAsync(Context context, ContentHashWithLastAc
                     return new PlaceFileResult(PlaceFileResult.ResultCode.NotPlacedAlreadyExists);
                 }
 
-                // If this is the empty hash, then put the empty file into the cache
-                if (_settings.UseEmptyFileHashShortcut && contentHashWithPath.Hash.IsEmptyHash() && !await ContainsAsync(context, contentHashWithPath.Hash, pinRequest))
+                // If this is the empty hash, then directly create an empty file.
+                // This avoids hash-level lock, all I/O in the cache directory, and even
+                // operations in the in-memory representation of the cache.
+                if (_settings.UseEmptyFileHashShortcut && contentHashWithPath.Hash.IsEmptyHash())
                 {
-                    // Put the empty file
-                    var putResult = await PutStreamAsync(context, _emptyFileStream, contentHash, pinRequest);
-                    if (!putResult.Succeeded)
-                    {
-                        // Because this is an optimization for the empty file, we shouldn't fail here.
-                        // Instead, let the normal path have a try.
-                        context.Debug($"Failed to place empty file by putting the empty stream: {putResult}");
-                    }
+                    // TODO: Respect I/O semaphore, set file permissions
+                    FileSystem.CreateDirectory(contentHashWithPath.Path.Parent);
+                    using (File.Create(contentHashWithPath.Path.Path)) { }

```

Comment:
I think we still should use FileSystem here, something like
`using(await FileSystem.CreateFile())`

where CreateFile is an extension method that calls OpenAsync with the right attributes.

I can see why this code will work, but I think its not a huge amount of work to use file system abstraction here.

Reply:
I don't think such a method exists on `IAbsFileSystem`, but it does exist on `BuildXL.Native.IO.FileUtilites` as `WriteAllBytesAsync` or `TryCreateOrOpenFile`.

---
In reply to: [279584408](https://github.com/Microsoft/BuildXL/pull/221#discussion_r279584408) [](ancestors = 279584408)

Reply:
I know, the idea is to add such an extension method that will do this.

---
In reply to: [279866313](https://github.com/Microsoft/BuildXL/pull/221#discussion_r279866313) [](ancestors = 279866313,279584408)

Reply:
You have convinced me. However, it's been such a pain getting checks to pass, I'm going to go ahead and merge now that they are passing, and will do a follow-up PR for this tomorrow.